In [1]:
import pandas as pd
import numpy as np

In [2]:
srv = pd.read_csv('immSurvey.csv')
srv.head()

Unnamed: 0   ids  X.1                                         textToSend  \
0           1  6606    1  problems caused by the influx of illegal immig...   
1           2  6607    2  if you mean illegal immigration, i'm afraid of...   
2           3  6608    3  that they should enter the same way my grandpa...   
3           4  6609    4  legally entering the usa meeting the requireme...   
4           5  6610    5         terror bombings killing us robbing america   

      Means  stanMeans  X  MetaID  treatment  pid_rep  ...    RunAll  \
0  0.750000   2.409768  1       0          1  1.00000  ...  0.750291   
1  0.875000   3.710615  2       0          1  1.00000  ...  1.011095   
2  0.416667  -1.437706  3       0          0  0.33300  ... -0.219371   
3  0.458333   0.655503  4       0          0  0.50000  ...  0.329845   
4  0.875000   5.337525  5       0          1  0.66667  ...  1.955240   

   mergeIDs  stanMeansNewSys  stanSDsNewSys  coderMeans  newDocIDS  \
0       240         1.919906       0.549770         1.5      10335   
1        75        -0.138688       0.302764         1.0      10336   
2       271        -0.334633       0.291336         0.0      10337   
3       181         0.386023       0.294324         1.5      10338   
4       261         0.102707       0.274519         2.0      10339   

   stanMeansNewSys2  stanSDsNewSys2  stanMeansNewSysPooled  \
0          1.106977        0.320538               1.465566   
1         -0.223681        0.300575              -0.166691   
2          0.825119        0.317955               0.224096   
3          0.679629        0.305606               0.629995   
4          1.206389        0.370577               0.574416   

   stanSDsNewSysPooled  
0             0.288474  
1             0.219645  
2             0.220390  
3             0.221096  
4             0.216228  

[5 rows x 38 columns]

In [3]:
alphas = srv.stanMeansNewSysPooled
sample = srv.textToSend

In [4]:
from sklearn.feature_extraction.text import CountVectorizer

In [5]:
vec = CountVectorizer()
X = vec.fit_transform(sample)
X

<340x1475 sparse matrix of type '<class 'numpy.int64'>'
	with 6239 stored elements in Compressed Sparse Row format>

In [6]:
pd.DataFrame(X.toarray(), columns=vec.get_feature_names())

11  12  125  18  1b  600  95  able  abolition  aboration  ...  wreckless  \
0     0   0    0   0   0    0   0     0          0          0  ...          0   
1     0   0    0   0   0    0   0     0          0          0  ...          0   
2     0   0    0   0   0    0   0     0          0          0  ...          0   
3     0   0    0   0   0    0   0     0          0          0  ...          0   
4     0   0    0   0   0    0   0     0          0          0  ...          0   
..   ..  ..  ...  ..  ..  ...  ..   ...        ...        ...  ...        ...   
335   0   0    0   0   0    0   0     0          0          0  ...          0   
336   0   0    0   0   0    0   0     0          0          0  ...          0   
337   0   0    0   0   0    0   0     0          0          0  ...          0   
338   0   0    0   0   0    0   0     0          0          0  ...          0   
339   0   0    0   0   0    0   0     0          0          0  ...          0   

     wrong  wrongly  xenophobia  years  yet  you  young  your  yrars  
0        0        0           0      0    0    0      0     0      0  
1        0        0           0      0    0    1      0     0      0  
2        0        0           0      0    0    0      0     0      0  
3        0        0           0      0    0    0      0     0      0  
4        0        0           0      0    0    0      0     0      0  
..     ...      ...         ...    ...  ...  ...    ...   ...    ...  
335      0        0           0      0    0    0      0     0      0  
336      0        0           1      0    0    0      0     0      0  
337      1        0           0      2    0    0      0     0      0  
338      0        0           0      0    0    0      0     0      0  
339      0        0           0      0    0    0      0     0      0  

[340 rows x 1475 columns]

In [7]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [8]:
vec = TfidfVectorizer()
X = vec.fit_transform(sample)
pd.DataFrame(X.toarray(), columns=vec.get_feature_names())

11   12  125   18   1b  600   95  able  abolition  aboration  ...  \
0    0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.0        0.0        0.0  ...   
1    0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.0        0.0        0.0  ...   
2    0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.0        0.0        0.0  ...   
3    0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.0        0.0        0.0  ...   
4    0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.0        0.0        0.0  ...   
..   ...  ...  ...  ...  ...  ...  ...   ...        ...        ...  ...   
335  0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.0        0.0        0.0  ...   
336  0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.0        0.0        0.0  ...   
337  0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.0        0.0        0.0  ...   
338  0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.0        0.0        0.0  ...   
339  0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.0        0.0        0.0  ...   

     wreckless     wrong  wrongly  xenophobia     years  yet       you  young  \
0          0.0  0.000000      0.0    0.000000  0.000000  0.0  0.000000    0.0   
1          0.0  0.000000      0.0    0.000000  0.000000  0.0  0.253915    0.0   
2          0.0  0.000000      0.0    0.000000  0.000000  0.0  0.000000    0.0   
3          0.0  0.000000      0.0    0.000000  0.000000  0.0  0.000000    0.0   
4          0.0  0.000000      0.0    0.000000  0.000000  0.0  0.000000    0.0   
..         ...       ...      ...         ...       ...  ...       ...    ...   
335        0.0  0.000000      0.0    0.000000  0.000000  0.0  0.000000    0.0   
336        0.0  0.000000      0.0    0.195271  0.000000  0.0  0.000000    0.0   
337        0.0  0.078393      0.0    0.000000  0.150361  0.0  0.000000    0.0   
338        0.0  0.000000      0.0    0.000000  0.000000  0.0  0.000000    0.0   
339        0.0  0.000000      0.0    0.000000  0.000000  0.0  0.000000    0.0   

     your  yrars  
0     0.0    0.0  
1     0.0    0.0  
2     0.0    0.0  
3     0.0    0.0  
4     0.0    0.0  
..    ...    ...  
335   0.0    0.0  
336   0.0    0.0  
337   0.0    0.0  
338   0.0    0.0  
339   0.0    0.0  

[340 rows x 1475 columns]

In [9]:
from sklearn.model_selection import train_test_split

In [10]:
Xtrain, Xtest, ytrain, ytest = train_test_split(X, alphas, random_state=1)

In [11]:
from sklearn.gaussian_process import GaussianProcessRegressor
from sklearn.gaussian_process.kernels import ConstantKernel, RBF

In [12]:
rbf = ConstantKernel(1.0)*RBF(length_scale=1.0)
gbr = GaussianProcessRegressor(kernel=rbf, alpha=1e-8)
gbr.fit(Xtrain.toarray(), ytrain)

C:\Users\user\anaconda3\lib\site-packages\sklearn\gaussian_process\_gpr.py:494: ConvergenceWarning: lbfgs failed to converge (status=2):
ABNORMAL_TERMINATION_IN_LNSRCH.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
  _check_optimize_result("lbfgs", opt_res)


GaussianProcessRegressor(alpha=1e-08, copy_X_train=True,
                         kernel=1**2 * RBF(length_scale=1),
                         n_restarts_optimizer=0, normalize_y=False,
                         optimizer='fmin_l_bfgs_b', random_state=None)

In [13]:
mu_s, cov_s = gbr.predict(Xtest.toarray(), return_cov=True)

In [14]:
np.corrcoef(ytest, mu_s)

array([[1.       , 0.6832852],
       [0.6832852, 1.       ]])

Correlation is 0.6833

BIGRAMS

In [15]:
bigram_vectorizer = CountVectorizer(ngram_range=(1, 2), token_pattern=r'\b\w+\b')

In [16]:
X = bigram_vectorizer.fit_transform(sample)

In [17]:
pd.DataFrame(X.toarray(), columns=bigram_vectorizer.get_feature_names())

1  1 3  1 became  1 boarders  1 difficult  1 immigrants  1 is  \
0    0    0         0           0            0             0     0   
1    0    0         0           0            0             0     0   
2    0    0         0           0            0             0     0   
3    0    0         0           0            0             0     0   
4    0    0         0           0            0             0     0   
..  ..  ...       ...         ...          ...           ...   ...   
335  0    0         0           0            0             0     0   
336  0    0         0           0            0             0     0   
337  0    0         0           0            0             0     0   
338  0    0         0           0            0             0     0   
339  0    0         0           0            0             0     0   

     1 language  1 not  11  ...  you think  you want  young  young from  your  \
0             0      0   0  ...          0         0      0           0     0   
1             0      0   0  ...          0         0      0           0     0   
2             0      0   0  ...          0         0      0           0     0   
3             0      0   0  ...          0         0      0           0     0   
4             0      0   0  ...          0         0      0           0     0   
..          ...    ...  ..  ...        ...       ...    ...         ...   ...   
335           0      0   0  ...          0         0      0           0     0   
336           0      0   0  ...          0         0      0           0     0   
337           0      0   0  ...          0         0      0           0     0   
338           0      0   0  ...          0         0      0           0     0   
339           0      0   0  ...          0         0      0           0     0   

     your poor  your tired  your weak  yrars  yrars ago  
0            0           0          0      0          0  
1            0           0          0      0          0  
2            0           0          0      0          0  
3            0           0          0      0          0  
4            0           0          0      0          0  
..         ...         ...        ...    ...        ...  
335          0           0          0      0          0  
336          0           0          0      0          0  
337          0           0          0      0          0  
338          0           0          0      0          0  
339          0           0          0      0          0  

[340 rows x 6703 columns]

In [18]:
Xtrain, Xtest, ytrain, ytest = train_test_split(X, alphas, random_state=1)

In [19]:
rbf = ConstantKernel(1.0)*RBF(length_scale=1.0)
gpr = GaussianProcessRegressor(kernel=rbf, alpha=1e-8)
gpr.fit(Xtrain.toarray(), ytrain)

GaussianProcessRegressor(alpha=1e-08, copy_X_train=True,
                         kernel=1**2 * RBF(length_scale=1),
                         n_restarts_optimizer=0, normalize_y=False,
                         optimizer='fmin_l_bfgs_b', random_state=None)

In [20]:
mu_s, cov_s = gpr.predict(Xtest.toarray(), return_cov=True)
np.corrcoef(ytest, mu_s)

array([[1.        , 0.60884464],
       [0.60884464, 1.        ]])

Correlation became 0.6088 when bigrams are included. It was 0.6833 without bigrams. 